File: benchmarks/python/run_all.py
#!/usr/bin/env python3
"""Run every Python benchmark sub-script in this directory and emit
`results/python.json` in the cross-binding schema.
"""

from __future__ import annotations

import json
import os
import platform
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)

import bench_persistence  # noqa: E402
import bench_primitives  # noqa: E402
import bench_recall  # noqa: E402


def main() -> int:
    print("Smritidb benchmark — Python binding")
    print(f"  Python {sys.version.split()[0]} on {platform.platform()}")

    primitives_out = bench_primitives.run()
    recall_out = bench_recall.run()
    persistence_out = bench_persistence.run()
    memory_out = bench_persistence.run_memory()

    output = {
        "binding": "python",
        "version": "0.1.0",
        "spec_version": primitives_out["spec_version"],
        "host": {
            "os": platform.system().lower(),
            "arch": platform.machine().lower(),
        },
        "primitives": primitives_out["results"],
        "recall": recall_out,
        "persistence": persistence_out,
        "memory": memory_out,
    }

    out_dir = os.path.abspath(os.path.join(HERE, "..", "results"))
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, "python.json")
    with open(out_path, "w") as f:
        json.dump(output, f, indent=2)
    print(f"wrote {out_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/typescript/run-all.ts
// Run every TypeScript benchmark, write results/typescript.json.

import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import os from "node:os";

import { SPEC_VERSION } from "@tanvrit/smritidb";

import { runPrimitives } from "./bench-primitives.js";
import { runRecall } from "./bench-recall.js";
import { runMemory, runPersistence } from "./bench-persistence.js";

const here = dirname(fileURLToPath(import.meta.url));

async function main(): Promise<number> {
  console.log("Smritidb benchmark — TypeScript binding");
  console.log(`  Node ${process.version} on ${os.platform()}/${os.arch()}`);

  const primitives = runPrimitives();
  const recall = runRecall();
  const persistence = await runPersistence();
  const memory = runMemory();

  const output = {
    binding: "typescript",
    version: "0.1.0",
    spec_version: SPEC_VERSION,
    host: { os: os.platform(), arch: os.arch() },
    primitives,
    recall,
    persistence,
    memory,
  };

  const resultsDir = resolve(here, "..", "results");
  mkdirSync(resultsDir, { recursive: true });
  const outPath = join(resultsDir, "typescript.json");
  writeFileSync(outPath, JSON.stringify(output, null, 2));
  console.log(`wrote ${outPath}`);
  return 0;
}

main().then((code) => process.exit(code));
